nx/lib/nx.ex

```diff
@@ -5176,15 +5176,114 @@ defmodule Nx do
   """
   @doc type: :indexed
   def indexed_add(target, indices, updates) do
+    impl = fn target, indices, updates, out ->
+      impl!(target, indices, updates).indexed_add(out, target, indices, updates)
+    end
+
+    indexed_op(target, indices, updates, impl)
+  end
+
+  @doc """
+  Performs an indexed `set` operation on the `target` tensor,
+  setting the `updates` into the corresponding `indices` positions.
+
+  `indices` must be a fully qualified tensor of shape `{n, Nx.rank(target)}`, with `n`
+  being an arbitrary number of indices, while `updates` must have a compatible `{n}` shape.
+
+  In case of repeating indices, the last occurence of index and its corresponding update
+  value wins.
+
+  ### Examples
+
+      iex> Nx.indexed_put(Nx.tensor([0, 0, 0]), Nx.tensor([[1], [2]]), Nx.tensor([2, 4]))
+      #Nx.Tensor<
+        s64[3]
+        [0, 2, 4]
+      >
+
+      iex> Nx.indexed_put(Nx.tensor([0, 0, 0]), Nx.tensor([[1], [2], [1]]), Nx.tensor([3, 4, 2]))
+      #Nx.Tensor<
+        s64[3]
+        [0, 2, 4]
+      >
+
+      iex> t = Nx.iota({1, 2, 3})
+      #Nx.Tensor<
+        s64[1][2][3]
+        [
+          [
+            [0, 1, 2],
+            [3, 4, 5]
+          ]
+        ]
+      >
+      iex> indices = Nx.tensor([[0, 0, 0], [0, 1, 1], [0, 0, 0], [0, 0, 2]])
+      iex> updates = Nx.tensor([1, 3, 2, -2])
+      iex> Nx.indexed_put(t, indices, updates)
+      #Nx.Tensor<
+        s64[1][2][3]
+        [
+          [
+            [2, 1, -2],
+            [3, 3, 5]
+          ]
+        ]
+      >
+
+  Type promotions should happen automatically.
+
+      iex> Nx.indexed_put(Nx.tensor([1.0]), Nx.tensor([[0]]), Nx.tensor([3]))
+      #Nx.Tensor<
+        f32[1]
+        [3.0]
+      >
+
+      iex> Nx.indexed_put(Nx.tensor([1]), Nx.tensor([[0]]), Nx.tensor([3.0]))
+      #Nx.Tensor<
+        f32[1]
+        [3.0]
+      >
+
+      iex> Nx.indexed_put(Nx.tensor([1], type: {:s, 32}), Nx.tensor([[0]]), Nx.tensor([3], type: {:s, 64}))
+      #Nx.Tensor<
+        s64[1]
+        [3]
+      >
+
+  ### Error cases
+      iex> Nx.indexed_put(Nx.tensor([[1], [2]]), Nx.tensor([[[1, 2, 3]]]), Nx.tensor([0]))
+      ** (ArgumentError) indices must be a rank 2 tensor, got: 3
+
+      iex> Nx.indexed_put(Nx.tensor([[1], [2]]), Nx.tensor([[1, 2]]), Nx.tensor([[0]]))
+      ** (ArgumentError) updates must be a rank 1 tensor, got: 2
+
+      iex> Nx.indexed_put(Nx.tensor([[1], [2]]), Nx.tensor([[1, 2, 3]]), Nx.tensor([0]))
+      ** (ArgumentError) expected indices to have shape {*, 2}, got: {1, 3}
+
+      iex> Nx.indexed_put(Nx.tensor([[1], [2]]), Nx.tensor([[1, 2]]), Nx.tensor([0, 1]))
+      ** (ArgumentError) expected updates tensor to match the first axis of indices tensor with shape {1, 2}, got {2}
+  """
+  @doc type: :indexed
+  def indexed_put(target, indices, updates) do
```

Comment:
For consistency with other functions :)

```suggestion
  def indexed_put(tensor, indices, updates) do
```

Reply:
I would keep it as `target` for consistency with `indexed_add` and because we're operating on 3 different tensors here

Reply:
Yeah, what I meant was to align `indexed_add` as well. If we want `target` then we should align `put_slice` as well, though I think `tensor` works for the leading argument, especially if it matches the wording in the docs. Sidenote: `take`/`take_along_axis`/`gather` use `tensor` instead of `source` :D

Reply:
I'd rather align other functions with `target`/`source` terminology, but I'm not set too hard on this

Reply:
Me neither, @josevalim feel free to opine on this :D

Reply:
I think that indexed_add, and indexed_put should retain the target terminology as it resolves any ambiguity in the difference between the 3 tensors.

On the other hand, for other functions, I don't think it makes any difference except the 3 tensors.

Reply:
I would go with `tensor` for consistency. indices and updates is enough to make it clear what it is.